src/OpenFeature/ProviderRepository.cs

```diff
@@ -0,0 +1,291 @@
+        /// <param name="context">the context to initialize the provider with</param>
+        /// <param name="afterSet">
+        /// <para>
+        /// Called after the provider is set, but before any actions are taken on it.
+        /// </para>
+        /// This can be used for tasks such as registering event handlers. It should be noted that this can be called
+        /// several times for a single provider. For instance registering a provider with multiple names or as the
+        /// default and named provider.
+        /// <para>
+        /// </para>
+        /// </param>
+        /// <param name="afterInitialization">
+        /// called after the provider has initialized successfully, only called if the provider needed initialization
+        /// </param>
+        /// <param name="afterError">
+        /// called if an error happens during the initialization of the provider, only called if the provider needed
+        /// initialization
+        /// </param>
+        /// <param name="afterShutdown">called after a provider is shutdown, can be used to remove event handlers</param>
+        public async Task SetProvider(string clientName,
+            FeatureProvider featureProvider,
+            EvaluationContext context,
+            Action<FeatureProvider> afterSet = null,
+            Action<FeatureProvider> afterInitialization = null,
+            Action<FeatureProvider, Exception> afterError = null,
+            Action<FeatureProvider> afterShutdown = null)
+        {
+            // Cannot set a provider for a null clientName.
+            if (clientName == null)
+            {
+                return;
+            }
+
+            this._providersLock.EnterWriteLock();
+
+            try
+            {
+                this._featureProviders.TryGetValue(clientName, out var oldProvider);
+                if (featureProvider != null)
+                {
+                    this._featureProviders.AddOrUpdate(clientName, featureProvider,
+                        (key, current) => featureProvider);
+                    afterSet?.Invoke(featureProvider);
+                }
+                else
+                {
+                    // If names of clients are programmatic, then setting the provider to null could result
+                    // in unbounded growth of the collection.
+                    this._featureProviders.TryRemove(clientName, out _);
+                }
+
+                // We want to allow shutdown to happen concurrently with initialization, and the caller to not
+                // wait for it.
+#pragma warning disable CS4014
+                this.ShutdownIfUnused(oldProvider, afterShutdown, afterError);
+#pragma warning restore CS4014
+            }
+            finally
+            {
+                this._providersLock.ExitWriteLock();
+            }
+
+            await InitProvider(featureProvider, context, afterInitialization, afterError).ConfigureAwait(false);
```

Comment:
At first I thought it was necessary to put this in the locked block, but I don't think it matters, since the important thing is to maintain thread safety of the provider dict, not make any guarantees about provider initialization order.